File: daily_python/498_Diagonal_Traverse.py
import os
import sys
from typing import List

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m = len(mat)
        n = len(mat[0])
        ans = []
        for round in range(m + n - 1):
            temp = []
            if round < n:
                i = 0
                j = round
            else:
                i = round - n + 1
                j = n - 1
            while i < m and j >= 0:
                temp.append(mat[i][j])
                i += 1
                j -= 1
            if round % 2 == 0:
                ans.extend(temp[::-1])
            else:
                ans.extend(temp)
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            mat = [list(map(int, row.split(','))) for row in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append(mat)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for mat in dataset:
        results.append(solution.findDiagonalOrder(mat))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/498_Diagonal_Traverse.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int m, n;
    int query(vector<vector<int>>& mat, int i, int j){
        if (i >= m || j >= n) return 1e9;
        return mat[i][j];
    }
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        m = mat.size();
        n = mat[0].size();
        vector<int> ans;
        ans.reserve(m * n);
        int round = 0;
        for (int i = 0; i < max(m, n); i++){
            vector<int> temp;
            for (int k = 0; k <= i; k++){
                int q = query(mat, i - k, k);
                if (q != 1e9) temp.push_back(q);
            }
            if (round++ % 2) reverse(temp.begin(), temp.end());
            for (int t : temp) ans.push_back(t);
        }
        for (int j = 1; j < max(m, n); j++){
            vector<int> temp;
            for (int k = 0; k < max(m, n) - j; k++){
                int q = query(mat, max(m, n) - 1 - k, j + k);
                if (q != 1e9) temp.push_back(q);
            }
            if (round++ % 2) reverse(temp.begin(), temp.end());
            for (int t : temp) ans.push_back(t);
        }
        return ans;
    }
};

int main(){
    vector<vector<int>> mat = {{1,2,3},{4,5,6},{7,8,9}};
    // vector<vector<int>> mat = {{1,2},{3,4}};
    // vector<vector<int>> mat = {{1,2,3,4},{5,6,7,8},{9,10,11,12},{13,14,15,16}};
    Solution S;
    vector<int> ans = S.findDiagonalOrder(mat);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}